### PR TITLE
Object level locking in Compliance and Governance modes CEPH-83574058

### PR DESCRIPTION
**Pull request by TejasC88**
Steps :
1. Create a bucket and enable locking , but do not set a lock policy on bucket
2. Create objects in bucket with the appropriate lock mode and retain until date.
3. Try to delete objects , should not work.

Pass logs :
http://magna002.ceph.redhat.com/ceph-qe-logs/tejas/object_level_retention_compliance.txt
http://magna002.ceph.redhat.com/ceph-qe-logs/tejas/object_level_retention_governance.txt

### CODE REVIEW

File: rgw/v2/tests/s3_swift/test_object_level_retention.py [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+from v2.tests.s3_swift import reusable
+from v2.tests.s3cmd.reusable import get_rgw_ip_and_port
+from v2.utils.log import configure_logging
+from v2.utils.test_desc import AddTestInfo
+from v2.utils.utils import RGWService
+
+log = logging.getLogger()
+TEST_DATA_PATH = None
+password = "32characterslongpassphraseneeded".encode("utf-8")
+encryption_key = hashlib.md5(password).hexdigest()
```
Comment (ckulal) [inline rgw/v2/tests/s3_swift/test_object_level_retention.py:49]: Please remove unused variable.

File: rgw/v2/tests/s3_swift/test_object_level_retention.py [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+                            s3_conn_client,
+                            bucket_name_to_create,
+                            s3_object_name,
+                            obj_body,
+                            lock_mode,
+                            retain_until,
+                        )
+
+                log.info("Try to delete the uploaded object versions")
+                versions = s3_conn_client.list_object_versions(
```
Comment (ckulal) [inline rgw/v2/tests/s3_swift/test_object_level_retention.py:107]: where do we handled uploading of version objects?

File: rgw/v2/tests/s3_swift/test_object_level_retention.py (hunk 1 above)
Comment (TejasC88) [reply rgw/v2/tests/s3_swift/test_object_level_retention.py]: Updated , thanks

File: rgw/v2/tests/s3_swift/test_object_level_retention.py (hunk 2 above)
Comment (TejasC88) [reply rgw/v2/tests/s3_swift/test_object_level_retention.py:105]: When we enable locking on bucket , the bucket becomes versioned by default

Comment (ckulal) [reply rgw/v2/tests/s3_swift/test_object_level_retention.py:105]: > When we enable locking on bucket , the bucket becomes versioned by default

As per current conf we are uploading only single version of object ryt..Can we have a conf for multiple version upload and test this scenario?

Comment (TejasC88) [reply rgw/v2/tests/s3_swift/test_object_level_retention.py:105]: actually as per current conf we are uploading single key and 30 versions of the same key

**Review by ckulal — COMMENTED** [review]
Looks good

**Review by anrao19 — APPROVED** [review]

**Review by viduship — DISMISSED** [review]

File: rgw/v2/tests/s3_swift/test_object_level_retention.py [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+                        ), "Expected: {expected_code}, Actual: {actual_code}"
+
+    # check for any crashes during the execution
+    crash_info = reusable.check_for_crash()
+    if crash_info:
+        raise TestExecError("ceph daemon crash found!")
+
+
+if __name__ == "__main__":
+    test_info = AddTestInfo("create m buckets with n objects")
```
[the remark this answers is not in the document]
Comment (TejasC88) [reply rgw/v2/tests/s3_swift/test_object_level_retention.py]: Updated , thanks @mkasturi18

**Review by mkasturi18 — APPROVED** [review]

**Pull request merged**